keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+
+There will be a hard break for clients and users of CRDs from `v1beta1` to the new `v1beta2`. We will provide a migration path for existing installed CRDs, but the clients accessing these CRDs will need to be updated.
+
+#### Conversion
+
+To migrate the existing CRDs, we will implement a Conversion Webhook that only allows `v1beta1` to `v1beta2` conversion, not the other way around. When KUDO is upgraded, it will:
+
+1. Deploy the new CRD version where the new (`v1beta2`) version will be the _only_ served and stored version:
+    - `v1beta1` will be still in the list, but with `storage=false, served=false`
+    - `v1beta2` will be marked as `storage=true, served=true`
+2. Deploy the new manager with the conversion webhook
+3. Run Pre-Migration: fetch all existing CRs, run the conversion and check for any errors
+4. Run the migration: fetch all CRDs and issue an update - This will trigger a conversion and save the CRD with the new version
+5. With a later major KUDO release, we can remove `v1beta1` from the CRD list - this should be at least 6 month later, as soon as we remove support for the `v1beta1` version we cannot support migrating existing installations anymore.
+
+Reference: [CRD Versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/)
+
+The actual type that we are going to use in the CRD is `extensionapi.JSON` - This type wraps a `[]byte` and allows storage of any data.  
+ 
+#### Operator
+
+The Operator CRD will not change, but we should keep the version in sync with the other CRDs.
+
+#### OperatorVersion
+
+OperatorVersion `v1beta1` [holds the parameters](https://github.com/kudobuilder/kudo/blob/main/pkg/apis/kudo/v1beta1/operatorversion_types.go#L35) in the `OperatorVersion.Spec.Parameter` field of type `[]Parameter`. `v1beta2` will update this field to hold a JSON-schema instead. This structure will be mapped in the CRD to a raw `extensionapi.JSON` and will be read and interpreted by a go library to e.g. validate input parameters and provide parameter defaults.
+
+`v1beta1`:
+```go
+type OperatorVersionSpec struct {
+    ....
+    Parameters []Parameter `json:"parameters,omitempty"`
+    ....
+}
+```
+
+`v1beta2`:
+```go
+type OperatorVersionSpec struct {
+    ....
+    Parameters extensionapi.JSON `json:"parameters"`
+    ....
+}
+```
+
+A better solution than `extensionapi.JSON` would be to have a typed structure, but CRDs currently cannot describe recursive data structures, as they [don't support $ref or $recursiveRef](https://github.com/kubernetes/kubernetes/issues/54579). The only way to define an arbitrary JSON structure in a Kubernetes API is with `apiextension.JSON`. The resulting CRD for the operator version will therefore look like this:
```

Comment:
Agreed. We will have to figure out how this strongly typed structure will look like - Maybe we can use something existing from a json-scheme go library, maybe we have to wrap that structure to make it easily usable.

Reply:
Well, I found this [json-schema validator](https://github.com/santhosh-tekuri/jsonschema) has this [type Schema struct](https://github.com/santhosh-tekuri/jsonschema/blob/master/schema.go#L20). But dragging this into the `OperatorVersion` ...

Reply:
Nice find 👍  We can't use that in the OV itself, because it won't get rendered in the CRD because of the missing capabilities for recursion. But we could maybe use it internally